### ISSUE TITLE
Weird warning recommending to use undocumented options

### ISSUE THREAD

**Issue by lwasyl** [detekt/detekt#3567]
## Expected Behavior
Build doesn't emit warnings or the warnings are **_easily_** actionable

## Observed Behavior
I see the following warning:
> 'failFast' is deprecated. Please use 'buildOnDefaultConfig' together with 'allRules'.

and it's not clear what to do about it.

## Steps to Reproduce
<details>
<summary>Detekt Gradle Plugin configuration:</summary>

```kotlin
configure<DetektExtension> {
    failFast = true // Enables all rules by default
    reports {
        xml.enabled = false
        html.enabled = false
    }
    input = files("src/")
    config = rootProject.files("quality/detekt.yml")
}
```
</details>

<details>
<summary>Detekt configuration file</summary>

```yaml
build:
  maxIssues: 0

processors:
  active: false

console-reports:
  active: true
  exclude:
    - 'ProjectStatisticsReport'
    - 'ComplexityReport'
    - 'NotificationReport'

comments:
  active: false

complexity:
  active: false

empty-blocks:
  active: true

exceptions:
  active: true
  NotImplementedDeclaration:
    active: false
  RethrowCaughtException:
    active: false
  ReturnFromFinally:
    active: false
  SwallowedException:
    active: false
  ThrowingExceptionsWithoutMessageOrCause:
    active: false
  ThrowingNewInstanceOfSameException:
    active: false
  TooGenericExceptionCaught:
    active: false
  TooGenericExceptionThrown:
    active: true
    exceptionNames:
      - Error
      - Exception
      - Throwable
      - RuntimeException

naming:
  active: true
  ForbiddenClassName:
    active: true
    forbiddenName: [ "^.*Impl$" ]
  FunctionMaxLength:
    active: false

performance:
  active: true
  SpreadOperator:
    active: false

potential-bugs:
  active: true
  CastToNullableType:
    active: false
  LateinitUsage:
    active: false
  UnsafeCast:
    active: false
  EqualsAlwaysReturnsTrueOrFalse:
    active: true

style:
  active: true
  CollapsibleIfStatements:
    active: false
  ClassOrdering:
    active: false
  DataClassContainsFunctions:
    active: false
  DataClassShouldBeImmutable:
    active: false
  DestructuringDeclarationWithTooManyEntries:
    excludes: [ '**/test/**' ]
  ExpressionBodySyntax:
    active: true
    includeLineWrapping: true
  ForbiddenComment:
    active: false
  FunctionOnlyReturningConstant:
    active: false
  LoopWithTooManyJumpStatements:
    active: false
  MagicNumber:
    active: false
  MaxLineLength:
    active: true
    maxLineLength: 140
  NestedClassesVisibility:
    active: false
  OptionalWhenBraces:
    active: false
  OptionalUnit:
    active: false # This generates quite a lot of false-positives
  PreferToOverPairSyntax:
    active: false
  ReturnCount:
    active: false
  SerialVersionUIDInSerializableClass:
    active: false
  ThrowsCount:
    active: false
  UnnecessaryAbstractClass:
    active: false
  UnusedPrivateClass:
    active: false
  UnusedPrivateMember:
    active: false
  UseDataClass:
    active: false
  UseIfInsteadOfWhen:
    active: false
```
</details>

## Context
Here are the steps I took to fix the warning: 

* I see the following warning:
> 'failFast' is deprecated. Please use 'buildOnDefaultConfig' together with 'allRules'.

* Googling `buildOnDefaulConfig` yields no reasonable results:
<img width="713" alt="image" src="https://user-images.githubusercontent.com/3951580/111465198-1079fd00-8722-11eb-949f-8b620cc246dd.png">

* Same thing on Detekt documentation page:
<img width="1192" alt="image" src="https://user-images.githubusercontent.com/3951580/111465162-07892b80-8722-11eb-9ea5-7cd837713084.png">

* No clear documentation in Google for `allRules`:
<img width="736" alt="image" src="https://user-images.githubusercontent.com/3951580/111465259-27b8ea80-8722-11eb-8080-e0f7d25e11b0.png">

* Same for Detekt documentation:
<img width="288" alt="image" src="https://user-images.githubusercontent.com/3951580/111465293-31425280-8722-11eb-9770-c2cfc051f3f4.png">

* I see no deprecation errors from the compiler so it takes me a while to realize that it's `failFast` in the Gradle plugin that's "deprecated", but not really deprecated: 
<img width="194" alt="image" src="https://user-images.githubusercontent.com/3951580/111465561-8e3e0880-8722-11eb-9812-0fa5cd2c962d.png">

_At this point I should note that in hindsight it's obvious that `failFast` comes from the gradle plugin. But also I have a vague recollection of `failFast` being in the detekt configuration file at one point and I was confused about the errors because I didn't see `failFast` option there. Detekt is just one part of our build and usually when encountering issues it's easier to google first than to dive into configuration thus the process._

* Replacing the `failFast` with `buildOnDefaultConfig` seems broken:
<img width="311" alt="image" src="https://user-images.githubusercontent.com/3951580/111465660-a9a91380-8722-11eb-862a-d5bf18aee96c.png">

* Fortunately IDE tells me that the property is in fact called differently
```diff
- buildOnDefaultConfig
+ buildUponDefaultConfig
```

* There is no documentation in the extension sources so I have no clue what this option does. Searching on Detekt documentation still doesn't work. Finally I find [the documentation for Gradle plugin](https://detekt.github.io/detekt/gradle.html) where I see some documentation in extremely inconvenient formatting:
<img width="877" alt="image" src="https://user-images.githubusercontent.com/3951580/111466087-1fad7a80-8723-11eb-92b5-e74b3ad72dfd.png">

```
// Interpret config files as updates to the default config. `false` by default.
```

* In the meantime I realized that `allRules` is also a configuration on the Gradle extension (which I initially missed because I forgot to sync the IDE). There's no javadoc for it either, search on documentation webpage still doesn't work, but this time there's also no mention of `allRules` on the page where I finally found `buildUponDefaultConfig` docs:
<img width="717" alt="image" src="https://user-images.githubusercontent.com/3951580/111466270-5b484480-8723-11eb-849f-30a8e3653ac3.png">

* I set both properties to `true` and hope everything works.

**_This was process of updating a minor version of the library by one_**, from previous most recent 1.15.0 to most recent 1.16.0. It took me _way_ too long.

As a general feedback, Detekt is usually the least convenient dependency to update. There are lots of moving parts, semver is not followed (there's no claim it is so that's fine). There are _**129**_ issues listed in the 1.16.0 changelog, 72 of which are in `changelog` section which, I assume, affects end users (as opposed to `Housekeeping & Refactorings` section). `Migration` section is empty, even though I just spent non-trivial amount of time figuring out how to _migrate_ away from a deprecated property. There's no mention of `buildOnDefaultConfig` or `buildUponDefaultConfig` in the changelog either. 

I have to add that I appreciate all the work put into Detekt, it's just as end user in the end I don't even know if after any given detekt bump the checks still work, and in a bigger project with non-trivial configuration it's not entirely obvious how the migration should look like if someone doesn't follow issues that caused the changes.

## Your Environment
* Version of detekt used: 1.16.0
* Version of Gradle used (if applicable): 6.8.3

**Comment by cortinico** [comment]
Thank you very much for your feedback @lwasyl.

I personally really appreciate receiving such detail feedback as it's a rare thing in the oss community, and we can use it to improve Detekt to support our users as best as we can.

To answer to your specific scenario, looks like a typo created a lot of confusion. We released 3 Release Candidates specifically to try to catch those bugs and problems. Sadly this time no one noticed/reported this typo and it ended up in production. It got fixed here #3572 afterwards.

As for the documentation being suboptimal, I agree. We greatly appreciate PRs that help us polish and improve our docs 🙏 The specific things you mentioned have been addressed here: #3596

> As a general feedback, Detekt is usually the least convenient dependency to update.

I hear you. As a general rule, we will try to release more often and with smaller releases. Releasing 1.16.0 was a big effort also for us, so we will try to make the process as lean as possible for everyone.

Also I agree that the CHANGELOG we provide is sometimes too long and not specifically useful. If we release smaller versions, we will try to provide a summary or an highlight section of the changes we're actually shipping.

> semver is not followed (there's no claim it is so that's fine).

Actually we do (the deprecation of the `failFast` property was specifically to follow semver). Have you noticed breaking changes between versions that impacted you? If yes, please report them.

Finally, we're all volunteers :) So we do the best that we can. Thanks for reporting this and I hope the update from 1.16.0 to 1.17.0 will be much smoother (PS: we're also on [#detekt](https://kotlinlang.slack.com/archives/C88E12QH4) on Kotlinlang if need further support on migrating or you have direct feedback to share).

**Comment by BraisGabin** [comment]
#3558 can be related.

And I agree about the changelog, we could try to find a way to make it more useful for our users.

**Comment by lwasyl** [comment]
@cortinico Thank you for answering 🙂 I realize that the core issue I had was just an unfortunate typo, and again I do appreciate all the work being put in Detekt and other open source libraries. Apologies if the report sounded a bit rant-y, I just wanted to report all the (relatively small) things I encountered along the way.

> As a general rule, we will try to release more often and with smaller releases. Releasing 1.16.0 was a big effort also for us, so we will try to make the process as lean as possible for everyone.

Honestly I wasn't aware 1.16 was a _big_ release. It's understandable that bugs happen, and I try to test pre-release versions as time permits, I realize reporting issues is crucial for the project to get better. 

> Also I agree that the CHANGELOG we provide is sometimes too long and not specifically useful. If we release smaller versions, we will try to provide a summary or an highlight section of the changes we're actually shipping.

Fun fact, I originally also reported that attaching changelog to Github release would be a nice addition, here https://github.com/detekt/detekt/issues/2150 🙂

> Actually we do (the deprecation of the failFast property was specifically to follow semver). Have you noticed breaking changes between versions that impacted you?

You're right, I didn't realize that deprecation is _not_ a breaking change, and I could've just let the deprecation warning be. I don't think I've had any actual breaking changes in the past releases

> PS: we're also on #detekt on Kotlinlang if need further support on migrating or you have direct feedback to share

I wasn't aware of that either, I'll be sure to head there in case of any questions 🙂

Thanks again!

**Comment by eygraber** [comment]
I came across this as well, but my question is why `allRules` or `buildUponDefaultConfig` are the suggested replacements for `failFast`? Shouldn't it be `ignoreFailures`?

**Comment by BraisGabin** [comment]
Why? Those two do different things. ignoreFailures doesn't active all the rules.

**Comment by eygraber** [comment]
I just looked up the old documentation for `failFast` and I think I get it now.

The doc for `ignoreFailures` says:

```
// If set to `true` the build does not fail when the
// maxIssues count was reached. Defaults to `false`.
ignoreFailures = false
```

When I see `failFast` it makes me think that the build will fail fast, i.e. it will fail at the first encountered error. So without any context as to what `failFast` was, I would think that it was the inverse of `ignoreFailures`.

**Comment by BraisGabin** [comment]
Yes, we deprecated it because the name was missleading.
